src/modules/workspaces/workspaces.controller.ts

```diff
@@ -458,15 +470,165 @@ export class WorkspacesController {
   @ApiOkResponse({
     description: 'Created File',
   })
-  @UseGuards(WorkspaceGuard)
+  @UseGuards(WorkspaceGuard, SharingPermissionsGuard)
   @WorkspaceRequiredAccess(AccessContext.WORKSPACE, WorkspaceRole.MEMBER)
+  @RequiredSharingPermissions(SharingActionName.UploadFile)
   async createFile(
     @Param('workspaceId', ValidateUUIDPipe)
     workspaceId: WorkspaceAttributes['id'],
+    @BehalfUserDecorator() behalfUser: User,
     @UserDecorator() user: User,
     @Body() createFileDto: CreateWorkspaceFileDto,
   ) {
-    return this.workspaceUseCases.createFile(user, workspaceId, createFileDto);
+    return this.workspaceUseCases.createFile(
+      behalfUser ?? user,
+      workspaceId,
+      createFileDto,
+    );
+  }
+
+  @Post('/:workspaceId/shared/')
+  @ApiOperation({
+    summary: 'Share file or folder to workspace',
+  })
+  @ApiBearerAuth()
+  @ApiParam({ name: 'workspaceId', type: String, required: true })
+  @ApiOkResponse({
+    description: 'Shared Item',
+  })
+  @UseGuards(WorkspaceGuard)
+  @WorkspaceRequiredAccess(AccessContext.WORKSPACE, WorkspaceRole.MEMBER)
+  async shareItemWithMember(
+    @Param('workspaceId', ValidateUUIDPipe)
+    workspaceId: WorkspaceAttributes['id'],
+    @UserDecorator() user: User,
+    @Body() shareItemWithTeam: ShareItemWithTeamDto,
+  ) {
+    return this.workspaceUseCases.shareItemWithTeam(
+      user,
+      workspaceId,
+      shareItemWithTeam,
+    );
+  }
+
+  @Get(':workspaceId/:teamId/shared/files')
+  @ApiOperation({
+    summary: 'Get shared files with a team',
+  })
+  @UseGuards(WorkspaceGuard)
+  @WorkspaceRequiredAccess(AccessContext.TEAM, WorkspaceRole.MEMBER)
+  async getSharedFiles(
+    @Param('teamId', ValidateUUIDPipe)
+    teamId: WorkspaceTeamAttributes['id'],
+    @UserDecorator() user: User,
+    @Query('orderBy') orderBy: OrderBy,
+    @Query('page') page = 0,
+    @Query('perPage') perPage = 50,
+  ) {
+    const order = orderBy
+      ? [orderBy.split(':') as [string, string]]
+      : undefined;
+
+    return this.sharingUseCases.getSharedFilesInWorkspaces(
+      user,
+      teamId,
+      page,
+      perPage,
+      order,
+    );
+  }
+
+  @Get(':workspaceId/:teamId/shared/folders')
+  @ApiOperation({
+    summary: 'Get shared folders with a team',
+  })
+  @UseGuards(WorkspaceGuard)
+  @WorkspaceRequiredAccess(AccessContext.TEAM, WorkspaceRole.MEMBER)
+  async getSharedFolders(
+    @Param('teamId', ValidateUUIDPipe)
+    teamId: WorkspaceTeamAttributes['id'],
+    @UserDecorator() user: User,
+    @Query('orderBy') orderBy: OrderBy,
+    @Query('page') page = 0,
+    @Query('perPage') perPage = 50,
+  ) {
+    const order = orderBy
+      ? [orderBy.split(':') as [string, string]]
+      : undefined;
+
+    return this.sharingUseCases.getSharedFoldersInWorkspace(
+      user,
+      teamId,
+      page,
+      perPage,
+      order,
+    );
+  }
+
+  @Get(':workspaceId/:teamId/shared/:sharedFolderId/folders')
```

Comment:
Following REST conventions, it needs to be like this:

GET /workspaces/:workspaceId/**teams**/:teamId/shared/:sharedFolderId/folders